openwisp_notifications/static/openwisp-notifications/js/notifications.js

```diff
@@ -11,9 +12,61 @@ if (typeof gettext === 'undefined') {
         notificationWidget($);
         initNotificationDropDown($);
         initWebSockets($);
+        notificationWindow.clearOldWindows();
+        notificationWindow.add();
     });
+
+    $(window).on('beforeunload', function () {notificationWindow.remove();});
 })(django.jQuery);
 
+const notificationWindow = {
+    add: function(){
+        // add window id to 'notificationWindows' in localStorage
+        let tempWindowsList = notificationWindow.list();
+        let index = tempWindowsList.indexOf(windowId);
+        if (index < 0){
+            tempWindowsList.push(windowId);
+        }
+        localStorage.notificationWindows = tempWindowsList.join(',');
+    },
+    remove: function(){
+        // remove window id from 'notificationWindows'
+        // in localStorage on reload and close
+        let tempWindowsList = notificationWindow.list();
+        let index = tempWindowsList.indexOf(windowId);
+        if (index > -1){
+            tempWindowsList.splice(index, 1);
+        }
+        localStorage.notificationWindows = tempWindowsList.join(',');
+    },
+    list: function(){
+        // returns 'notificationWindows' array
+        if (localStorage.notificationWindows){
+            return localStorage.notificationWindows.split(',');
+        }
+        return [];
+    },
+    canPlaySound: function(){
+        // returns true if the current window is
+        // the last one saved in localStorage
+        let windowList = notificationWindow.list();
+        let lastWindow = windowList && windowList[windowList.length-1];
+        return lastWindow && lastWindow === windowId;
+    },
+    clearOldWindows: function(){
+        // remove windows older than 12 hours
+        let tempWindowsList = notificationWindow.list();
+        let newWindowsList = tempWindowsList.map((window)=>{
+            if(((+windowId)-(+window))/36e5<=12){
+                return window;
+            }
+        });
+        localStorage.notificationWindows = newWindowsList.join(',');        
+    }
+};
```

Comment:
```suggestion
const oWNotificationWindow = {
    set: function(){
        // Sets the last focused window
        localStorage.setItem('owLastWindow', owWindowId);
    },
    remove: function () {
        // Removes the last focused window entry if current window
        // was referenced
        if (localStorage.getItem('owLastWindow') === owWindowId){
            localStorage.removeItem('owLastWindow');
        }
    },
    canPlaySound: function(){
        // Returns true if the current window was the last fouces one
        return localStorage.getItem('owLastWindow') === owWindowId;
    },
};
```

I tried tried this implementation where we never store more than one `windowId` in `localStorage` at any time. This implementation has following benefits:

- It is quicker. We are not dealing with array, we are reducing time and space complexity. 
- It does not bloats `localStorage` as concerned by @nemesisdesign. At any time, there will be atmost one instance of the last focused window. Even if the value is not removed while closing a tab, it will get overridden when a new window is opened or focus in changed to another window.

I didn't understand to keep track of all opened windows in the first place. I think the problem is allow sound to be played only from one window only, We need not worry about how many windows were opened by the user. @niteshsinha17 do you still think we need to maintain the list of opened windows?

Reply:
The purpose of storing old windows was the following reasons:

- To change the authority of playing sound if needed.

- Suppose if all opened windows of openwisp-notifications are not focused and the user suddenly closes the last window during being on some other website then the authority should be given to 2nd last or another window to insure that sound plays when a notification is received. It's not possible if we don't store all windows id. 

- Suppose two windows of the same browser are opened not on tab different tabs but on two different windows then it is possible to focus on both windows at the same time by minimizing the window size. Then which window will play sound in this situation. 

User can do anything, I have tried to target all the possibilities that a user can do. The Idea you suggested was also in my mind but because of these problems, I thought to store window id. It definitely increases time and space complexity but the size of the array is not very large.

Reply:
> Suppose if all opened windows of openwisp-notifications are not focused and the user suddenly closes the last window during being on some other website then the authority should be given to 2nd last or another window to insure that sound plays when a notification is received. It's not possible if we don't store all windows id.

Good observation. In that case, we can just add another event listener to handle it. Whenever there is a change in localStorage, `storage.onChange` event is fired. We can handle this event to assign the id of the current window to the localStorage. I acknowledge this will create a race condition, but since in this case, we can play the sound from any window randomly that should not be a concern.

```

   $(window).on('storage', function () {
        if (localStorage.getItem('owLastWindow') === null) {
            oWNotificationWindow.set();
        }
    });
```

> Suppose two windows of the same browser are opened not on tab different tabs but on two different windows then it is possible to focus on both windows at the same time by minimizing the window size. Then which window will play sound in this situation.

Did you test this? I tested this scenario, but the sound was getting played from only one window. Focus is something different from which window is currently occupying the screen. Can you confirm whether you are still able to reproduce it? If you do, share the steps you took to reproduce this. 

> User can do anything, I have tried to target all the possibilities that a user can do. 

I agree, we should try to cover all the cases. But, we can give up on few rare ones like "what if the notification arrives during the execution of storage event handler" to avoid bloating the code.

Reply:
@pandafy from the suggested change it's not immediately evident to me what happens when the user closes the window which plays the sound? How is another windows chosen to play the sound?

Reply:
> Good observation. In that case, we can just add another event listener to handle it. Whenever there is a change in localStorage, storage.onChange event is fired. We can handle this event to assign the id of the current window to the localStorage. I acknowledge this will create a race condition, but since in this case, we can play the sound from any window randomly that should not be a concern.

Here you are calling `set` inside the change handler, It will start an infinite loop which will change `window id` infinite times. This doesn't look good.

Reply:
> Did you test this? I tested this scenario, but the sound was getting played from only one window. Focus is something different from which window is currently occupying the screen. Can you confirm whether you are still able to reproduce it? If you do, share the steps you took to reproduce this.

Sorry, I hadn't tested it. I was thinking focus is something related to visibility and I was wrong.

Reply:
I have one more solution which doesn't have any race condition or doesn't store all window id.
In this, I am storing window Id as 0,1,2,3... Code will become as follows:
```
const owNotificationWindow = {
    set: function(){
        // set window id to 'notificationWindows' in localStorage
        // and returns the window id
        localStorage.setItem('owWindow',+localStorage.getItem('owWindow')+1);
        return +localStorage.getItem('owWindow');
    },
    remove: function(){
        // remove window id from 'notificationWindows'
        // in localStorage on reload and close
        localStorage.setItem('owWindow',+localStorage.getItem('owWindow')-1);
    },
    canPlaySound: function(){
        // returns true if the current window is
        // same as saved in localStorage
        return +localStorage.getItem('owWindow') === owWindowId;
    },
    changed: function(){
        // change window id if window id in localStorage
        // is greater than this window id
        if(owWindowId > +localStorage.getItem('owWindow')){
            owWindowId-=1;
        }
    }
};
var owWindowId = owNotificationWindow.set();
```
```
(function ($) {
    $(document).ready(function () {
        notificationWidget($);
        initNotificationDropDown($);
        initWebSockets($);
    });

    $(window).on('beforeunload', function () {owNotificationWindow.remove();});
    $(window).on('storage', function () {
        owNotificationWindow.changed();
    });
})(django.jQuery);
```

```
if (owNotificationWindow.canPlaySound()){
                // prevent playing of sound from multiple window
                notificationSound.currentTime = 0;
                notificationSound.play();
            }
```
In this solution, we need to use var in the global variable. Please help to find the best one according to you.

Reply:
> Here you are calling set inside the change handler, It will start an infinite loop which will change window id infinite times. This doesn't look good.

@niteshsinha17  again, did you test it? If this is the case, share the steps you took to reproduce it.

I could not find the link earlier but here is the [W3 specification](https://www.w3.org/TR/2008/WD-html5-20080610/structured.html) for firing `Storage.onChange` event

> When the setItem(), removeItem(), and clear() methods are called on a Storage object x that is associated with a local storage area, then in every HTMLDocument object whose Window object's localStorage attribute's Storage object is associated with the same storage area, other than x, a storage event must be fired, as described below.

Look for it in section 4.11.1.4. 

> @pandafy from the suggested change it's not immediately evident to me what happens when the user closes the window which plays the sound? How is another windows chosen to play the sound?

@nemesisdesign the new event handler will manage this case

> Suppose if all opened windows of openwisp-notifications are not focused and the user suddenly closes the last window during being on some other website then the authority should be given to 2nd last or another window to insure that sound plays when a notification is received. It's not possible if we don't store all windows id.

When the window is closed in this scenario, the `Storage.onChange` event will be fired in other windows. The authority of playing sound will be given to either the first one to complete the execution or the last one to set key in localStorage.

Reply:
> @niteshsinha17 again, did you test it? If this is the case, share the steps you took to reproduce it.
@pandafy  again sorry, I not noticed it `localStorage.getItem('owLastWindow') === null`. I will not do it again. According to me solution you have given is the best. Can I do changes as you have suggested?

Reply:
> @niteshsinha17 again, did you test it? If this is the case, share the steps you took to reproduce it.


@pandafy again sorry, I not noticed it localStorage.getItem('owLastWindow') === null. I will not do it again. According to me solution you have given is the best. Can I do changes as you have suggested?

Reply:
> @niteshsinha17 again, did you test it? If this is the case, share the steps you took to reproduce it.

@pandafy again sorry, I haven't noticed it `localStorage.getItem('owLastWindow') === null`. Its not good. I will try to avoid it next time. 

I like your solution. Can I do changes as you suggested?